WordPress/gutenberg
# UI Components: Refactor BlockSettingsMenu and MoreMenu as DropdownMenu

Previously: https://github.com/WordPress/gutenberg/pull/13410#discussion_r272586986 , #2989
Related: #14754 
Supersedes: #13059
Fixes #12505, fixes #15318 (Single visible menu is made up of multiple menus)

This pull request seeks to refactor the block settings menu to use the [`DropdownMenu` component](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown-menu). This is part of a larger effort to consolidate and normalize behaviors between dropdown menus. Prior to this pull request, there was a fair amount of duplication and divergent behaviors between the block settings menu and the standard `DropdownMenu` component.

There is not expected to be a difference in the look and feel of the block settings menu.

**Design Notes:**

The changes here err on the side of treating the block settings menu styles as the more correct iteration of dropdown menu styling. Notably, this has the following impact:

- The right margin of an icon in DropdownMenu has increased by `1px`, from `4px` to `5px`
- Default left- and right-padding of the DropdownMenu contents has been removed. This affects existing usage in the table block Edit Table dropdown ([before](https://user-images.githubusercontent.com/1779930/55639585-b92e2c80-5797-11e9-9abf-2ecdce0db1c2.png), [after](https://user-images.githubusercontent.com/1779930/55639502-86843400-5797-11e9-9821-22f54e7e8961.png)).
- Small viewport hover styling has been restored (previously removed by #10333 in an attempt to remove tap-drag hover activation described in #10320). Again noting that this never applied previously for block settings menu, but did apply for "More Options" and "Edit Table" dropdowns. This may need design confirmation. However, I would argue that it is not safe to assume that the relationship between touch device and small viewports are mutually inclusive (e.g. laptops with touchscreen support). 

**Implementation Notes:**

I operated on an assumption that the fact that block settings menus and "More Options" menu were not implemented as `DropdownMenu` was at least partly motivated by the amount of additional customization necessary for these menus being unsupportable by the base component. It's for this reason that I introduced a new `children` render function prop to serve as an alternative to the existing (limiting) `controls` prop.

As an alternative, I had also explored whether to allow elements to be intermixed with `control` object definitions in the default `controls` prop. There is a possible route here, but it is made difficult by the fact that (a) the menu items are likely to need access to `onClose`, (b) each element in an array needing a `key`, and (c) the confusion surrounding an overloaded array supports of an object and element formats. It's for this reason that I decided to stick with the render function and, if anything, I would suggest that in retrospect this makes more sense as a default anyways.

It may help to review using [GitHub's "Ignore white space changes" option](https://github.blog/2018-05-01-ignore-white-space-in-code-review/) (especially `BlockSettingsMenu`).

**Testing Instructions:**

Verify that the block settings menu ([right-most in toolbar](https://user-images.githubusercontent.com/1779930/55640179-18407100-5799-11e9-8ca5-5f36d142c607.png)) looks and behaves effectively identical as to `master`.

Verify that More Menu ([right-most in header](https://user-images.githubusercontent.com/699132/56042574-aab5b700-5d3b-11e9-8c76-0f6df6df3e2b.png)) looks and behaves identically as to `master` with one exception:
- using up/down arrow keys inside the menu allows you to navigate between all items, it doesn't lock you down inside a single group as proposed in #12505

![more-menu](https://user-images.githubusercontent.com/699132/56042662-e18bcd00-5d3b-11e9-99b1-0de6904c3075.gif)


Ensure unit tests pass:

```
npm run test-unit
```

**Included Tasks:**
- Refactor "More Options" as `DropdownMenu`
- Refactor `MenuGroup` to stop using `NavigableMenu` which wrongly narrows down arrow keys navigation

**Known bugs**:

There are 2 bugs which are being tackled separately:

- Focus wrongly transitions to post title when toggling Visual/Code (#14850)
- More Options menu toggle buttons cause focus loss when toggled (#14849)

**Future Tasks:**

- Refactor `DropdownMenu` to use `MenuItem` component in its default rendering of `controls`
